### PR TITLE
Test reboot

### LINKED ISSUE #7412: The external IP field for all nodes on GCE contains the master's IP address (instead of the node address)
When we populate the addresses of the nodes in [nodecontroller.go](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/nodecontroller/nodecontroller.go#L283) we call NodeAddresses, which for the gce provider gets the external IP from the [metadata server](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/cloudprovider/gce/gce.go#L417) which as expected puts the IP address of the master into the external IP field. 

The Legacy IP field is still filled in by asking the cloud provider for the instance data, so it contains the public IP of the node. 

/cc @mbforbes

### PR DESCRIPTION
**Pull request by mbforbes**
- It tests rebooting nodes, checking node and pod readiness before and after.
- It also includes a minor SSH refactoring to make the SSH test only the driving logic (special libraries all in util).

### CODE REVIEW

**Comment by mbforbes** [comment]
+cc @zmerlynn @ixdy @quinton-hoole

**Comment by mbforbes** [comment]
FYI, this test takes ~80-90 seconds to run.

(That is on a two-node cluster. However, the node and pod checking happens in parallel, so presumably it wouldn't increase significantly for more nodes, except for capturing a wider sampling of the natural timing variance.)

**Comment by zmerlynn** [comment]
I'm surprised it's only 90s. I've seen reboots take as long as 120s to get
happy. But still manageable.

**Comment by quinton-hoole** [comment]
80-90 sec is a bit longer than ideal, but I think it's more important right now that the test exists than that it is fast.  Given that we should be able to run tests in parallel pretty soon, the running time of individual tests becomes less important.  This one will need it's own cluster to be able to run in parallel though.

**Comment by roberthbailey** [comment]
Let me know when you've re-pushed and want me to review again.

**Comment by mbforbes** [comment]
Thanks for the review! I've addressed all the comments and re-pushed. I'm running the complete e2e's now (which will really be the interesting case if this breaks, for example, volumes).

@quinton-hoole Any ideas how to get this test to run faster? I'm doing all the reboots in parallel and the majority of the time is just waiting for the node to go down / up.

**Comment by roberthbailey** [comment]
Shippable and Travis both seem broken (looks like gofmt errors). Can you verify that the tests pass for you locally?

**Comment by zmerlynn** [comment]
You won't get it to run faster.

**Comment by zmerlynn** [comment]
(We can make reboot optimizations.. The hashes I put up in #7602 will help
minorly with a configure-vm change, as will an internal containerVM change,
but this test isn't getting faster on its own.)

You won't get it to run faster.

**Comment by mbforbes** [comment]
Ugh.

- `for range` is valid syntax in Go 1.4. If you write `for _ := range` gofmt tells you to simplify it.
- `for range` is not valid syntax in Go 1.3. It's OK with `for _ := range`, though.

**Comment by mbforbes** [comment]
... and neither 1.3 nor 1.4 is happy with `for _ := range len(...)`.

Silly question: why are we testing on both Go 1.3 and 1.4? It's looking like I'm going to have to add an unnecessary variable and some logging that uses it to get around this, unless I'm being obviously crazy about this?

**Comment by mbforbes** [comment]
Well, I added some unnecessary logging so I could use the old `range` syntax and have Go 1.4 be happy about it. I added a TODO to remove this once we stop supporting Go < 1.4.

Regarding e2es, I'll launch off another full run now. When I ran them earlier this morning, the reboot test failed due to a timeout on one of the nodes being ready again after reboot, so I upped this from 2 to 3 minutes (though unsure if this could have been because I closed my laptop at some point). Also, the monitoring test failed, so seeing whether that happens again.

@zmerlynn Good to know I'm not missing any speed tricks.

**Comment by zmerlynn** [comment]
Honestly, I'm not comfortable with 3m as a hard timeout - remember that a timeout needs to hit the 99% case, and I'm not sure 3m is 99th percentile due to the stupid stuff configure-vm still has to do on reboot. I would be more comfortable with 5m timeout when this first goes in, and we can track it.

**Comment by quinton-hoole** [comment]
@mbforbes I agree with zmerlynn@, it's going to be difficult or impossible to get it to run much faster.

**Comment by roberthbailey** [comment]
Re the range syntax, @a-robinson hit this recently as well.

**Comment by a-robinson** [comment]
Heh, yeah I hit the 1.3 vs 1.4 range thing as well. If you really don't want the logging, you can just do an empty assignment on the next line to make go stop complaining:
```
for foo := range bar {
  _ = foo
  ...
}
```

**Comment by mbforbes** [comment]
I upped the "node ready again" timeout to 5 minutes.

This had been failing monitoring e2es with the error:

`INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative`

which may have been because of drift of my cluster code and test code. I'm spinning up a new e2e cluster from head and will run the full e2e test suite on it.

**Comment by roberthbailey** [comment]
@mbforbes how did your e2e run go? I'd rather not break e2es over the weekend so unless you have a successful run I'm planning to wait to merge until monday.

File: test/e2e/reboot.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	It("should reboot each node and ensure they function upon restart", func() {
+		// This test requires SSH, so the provider check should be identical to
+		// there (the limiting factor is the implementation of util.go's
+		// getSigner(...)).
+		provider := testContext.Provider
+		if !providerIs("gce", "gke") {
```
Comment (roberthbailey) [inline test/e2e/reboot.go]: If we only enable this for gce is it ready to go in?

**Comment by mbforbes** [comment]
Running e2e again now that Jenkins is relatively green, it looks like there are three legitimate test failures, all of which happened after the reboot test ran:

```bash
Summarizing 3 Failures:

[Fail] kubectl guestbook [It] should create and stop a working application
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/kubectl.go:133

[Fail] Density [AfterEach] [Performance suite] should allow starting 3 pods per node
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/density.go:84

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:337

Ran 40 of 46 Specs in 2761.502 seconds
FAIL! -- 37 Passed | 3 Failed | 1 Pending | 5 Skipped
```

I'll post the logs from each failure in a gist in a followup comment, and will start investigating them now; help is welcome, of course, so just ping on this thread if you're looking into one of them.

Question: is there an easy way to tell ginkgo to run a test first without playing with seeds? I'd like to run the reboot test right off the bat, because it could be there are other tests that only fail after the reboot has happened. I guess I could just disable the reboot test for now and run e2es again and see what breaks. I'll try that now.

**Comment by mbforbes** [comment]
Here are the logs for the three failed tests:

0. [guestbook](https://gist.github.com/mbforbes/ff74c87346f724b0fc8e)
0. [density](https://gist.github.com/mbforbes/99f9506b83b82da476da)
0. [Elasticsearch](https://gist.github.com/mbforbes/7e41e970229482fe4569)

**Comment by roberthbailey** [comment]
@ixdy @quinton-hoole

**Comment by roberthbailey** [comment]
Were those the only tests than ran after reboot or were there other ones that passed?

**Comment by mbforbes** [comment]
There were other tests that ran after the reboot and passed. I'm running again to see if anything else fails after a reboot happens (I tried using `--test_args='--ginkgo.focus=^((?!Reboot).)*$'` to run all but the reboot test, but ginkgo wasn't happy about this).

**Comment by mbforbes** [comment]
It's looking a bit more than trivial to debug these failures, so I'm going to split into separate issues. I think understanding why these tests fail will help us understand what we need to think about re: node state for upgrades. I'll update this comment and the one above in the next few minutes as I create issues for each.

**Comment by roberthbailey** [comment]
@mbforbes Poking this issue -- fyi it needs a rebase.

**Comment by mbforbes** [comment]
Yeah I've since rebased and have been running e2e tests. I'll push again now.

I'm getting a monitoring failure now:

```
[Fail] Monitoring [It] verify monitoring pods and all cluster nodes are available on influxdb using heapster.
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:236
```

... I'm not sure whether it's just a naming problem?


```
Monitoring
  verify monitoring pods and all cluster nodes are available on influxdb using heapster.
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51
[BeforeEach] Monitoring
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:42
>>> testContext.KubeConfig: /Users/maxforbes/.kube/config
[It] verify monitoring pods and all cluster nodes are available on influxdb using heapster.
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: expected only one series from Influxdb for query "select distinct(pod_id) from /cpu.*/". Got []
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative
INFO: failed to query list of pods from influxdb. Query: "select distinct(pod_id) from /cpu.*/", Err: Server returned (400): Field pod_id doesn't exist in series cpu/usage_ns_cumulative

• Failure [303.125 seconds]
Monitoring
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:52
  verify monitoring pods and all cluster nodes are available on influxdb using heapster. [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:51

  monitoring using heapster and influxdb test failed

  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/monitoring.go:236
```

**Comment by roberthbailey** [comment]
@vishh @vmarmol

**Comment by mbforbes** [comment]
@vishh helped me debug; may be a kubelet bug; filing an issue...

**Comment by dchen1107** [comment]
FYI: PRs to fix the race during the initialization between docker and cadvisor are merged.

**Comment by mbforbes** [comment]
Cool, running again now. Hopefully e2es are behaving today. Thanks for all the help!

**Comment by mbforbes** [comment]
Just an FYI to anyone still reading along, I've noticed a marked regression in the timing this test takes since I've been running it. When I first introduced it, it would consistently take ~1.5 minutes. Now, it consistently takes ~3.5 minutes. I noticed this certainly after #8191 was fixed, though possibly before. It could be a result of legitimate fixes (e.g. we now correctly wait for more things on reboot and avoid races), but I just wanted to note it here.

**Comment by mbforbes** [comment]
Removed "gke" as a provider for now as "gce" is passing for sure. Kicked shippable as #8357 is closed. Will ping on green—looking forward to getting this in!

File: test/e2e/reboot.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		if !<-result {
+			Logf("Pod %s failed to be %s.", podName, desc)
+			success = false
+		}
+	}
+	Logf("Wanted all pods to be %s. Result: %t. Pods: %v", desc, success, podNames)
+	return success
+}
+
+// waitForNodeToBe returns whether node name is ready within timeout.
```
Comment (roberthbailey) [inline test/e2e/reboot.go]: nit: function comment should be waitForNodeToBeReady

File: test/e2e/reboot.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+func waitForNodeToBeReady(c *client.Client, name string, timeout time.Duration) bool {
+	return waitForNodeToBe(c, name, true, timeout)
+}
+
+// waitForNodeToBe returns whether node name is not ready (i.e. the readiness
```
Comment (roberthbailey) [inline test/e2e/reboot.go]: nit: function comment should be waitForNodeToBeNotReady

File: test/e2e/reboot.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (roberthbailey) [inline test/e2e/reboot.go]: 2015

Comment (mbforbes) [inline test/e2e/reboot.go]: Done.

File: test/e2e/reboot.go (hunk 3 above)
Comment (mbforbes) [inline test/e2e/reboot.go]: Done.

File: test/e2e/reboot.go (hunk 2 above)
Comment (mbforbes) [inline test/e2e/reboot.go]: Done.

**Comment by roberthbailey** [comment]
Shippable is happy. Merging.

**Comment by mbforbes** [comment]
Woohoo!

**Pull request merged**